task/dgm.py

```diff
@@ -1,171 +1,360 @@
+        })
+
+        if len(self.dict_list) >= self.transaction_size:
+            self.write_batch()
+
+        return self.transaction_size - len(self.dict_list)
+
+    def flush_dict(self):
+        self.write_batch()
+
+    def read_batch(self, collection_id, provider_path, batch_size):
+        batch = self.model.read_batch(collection_id, provider_path, batch_size)
+        self.queued_files_count += len(batch)
+        return batch
+
+    def write_batch(self):
+        self.discovered_granules_count += getattr(self.model, f'db_{self.duplicate_handling}')(self.dict_list)
+        self.dict_list.clear()
+
+
+class DBManagerCumulus(DBManagerBase):
+    def __init__(self, db_type, duplicate_handling, transaction_size):
+        super().__init__(db_type, duplicate_handling, transaction_size)
+        self.DB = psycopg2.connect(**db_init_kwargs) if 'psycopg2' in globals() else None
 
-        return del_count
+    def close_db(self):
+        self.DB.close()
 
-    def fetch_batch(self, collection_id, provider_path, batch_size=1000, **kwargs):
+    def add_record(self, name, granule_id, collection_id, etag, last_modified, size):
+        self.dict_list.append({
+            'name': name,
+            'etag': etag,
+            'granule_id': granule_id,
+            'collection_id': collection_id,
+            'last_modified': str(last_modified),
+            'size': size
+        })
+
+    def flush_dict(self):
+        if self.duplicate_handling == 'skip':
+            db_granule_ids = self.trim_results()
+            db_granule_ids = set(db_granule_ids)
+            print(f'Trimming {len(db_granule_ids)} files that already existed in the Cumulus database.')
+            print(f'Trimmed granule IDs: {db_granule_ids}')
+
+            # Remove the keys that have already been discovered
+            index = 0
+            while index < len(self.dict_list):
+                granule_id = self.dict_list[index].get('granule_id')
+                if granule_id in db_granule_ids:
+                    del self.dict_list[index]
+                else:
+                    index += 1
+
+        self.discovered_granules_count += len(self.dict_list)
+
+    def read_batch(self, collection_id, provider_path, batch_size):
+        self.queued_files_count += len(self.dict_list)
+        return self.dict_list
+
+    def trim_results(self):
+        granule_ids = [x.get('granule_id') for x in self.dict_list]
+        print(f'granule_ids: {granule_ids}')
+        results = []
+        start_index = 0
+        end_index = VAR_LIMIT
+        db_st = time.time()
+        while True:
+            with self.DB:
+                with self.DB.cursor() as curs:
+                    id_batch = tuple(granule_ids[start_index:end_index])
+                    if len(id_batch) == 0:
+                        break
+                    print(f'id_batch" {id_batch}')
+                    query_string = 'SELECT granules.granule_id FROM granules WHERE granules.granule_id IN %s;'
```

Comment:
I have discussed how this could lead to problems if unbounded queries tie up the database (e.g., asking for 3 million records is going to be really bad) with Mike. I learned that this logic is only reachable when following the Cumulus Database deployment option, which is incompatible with our expressed end-goals regardless. Mike has also, additionally, vowed to dedicate some time to pondering testing or exploring this theoretical issue more in-depth.